File: algorithms/538. Convert BST to Greater Tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        def dfs(root):
            if root:
                dfs(root.right)
                root.val = total[0] = total[0] + root.val
                dfs(root.left)
        total = [0]
        dfs(root)
        return root

File: algorithms/538. Convert BST to Greater Tree/solution.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    int total = 0;
public:
    TreeNode* convertBST(TreeNode* root) {
        dfs(root);
        return root;
    }

    void dfs(TreeNode * root) {
        if ( root ) {
            dfs(root->right);
            total = root->val = total + root->val;
            dfs(root->left);
        }
    }
};
